File: src/main/java/com/elpassion/whyrx/ReactiveExample.java
package com.elpassion.whyrx;

import java.util.concurrent.Executor;

public class ReactiveExample {

    private final Executor executor;

    public ReactiveExample(Executor executor) {
        this.executor = executor;
    }

    public Observabe<Double> calculate(double input) {
        return Observabe.just(input)
                .map(this::log10)
                .map(this::log10)
                .subscribeOn(executor);
    }

    private Double log10(Double input) {
        double log10 = Math.log10(input);
        if (Double.isNaN(log10)) {
            throw new IllegalArgumentException();
        }
        return log10;
    }
}

File: src/main/java/com/elpassion/whyrx/ReactiveExample.kt
package com.elpassion.whyrx

import java.util.concurrent.Executor

class ReactiveExample(val executor: Executor) {

    fun calculate(input: Double): Observable<Double> {
        return Observable(executor, Provider { calculateLog10(input) })
                .flatMap({ calculateLog10(it) })
    }

    private fun calculateLog10(input: Double): Double {
        val log10 = log10(input)
        if (log10.isNaN()) {
            throw RuntimeException("Illegal argument")
        }
        return log10
    }
}